processing/src/main/java/io/druid/query/groupby/orderby/DefaultLimitSpec.java

```diff
@@ -320,6 +343,78 @@ public int hashCode()
     }
   }
 
+  public class GroupTopNFunction implements Function<Sequence<Row>, Sequence<Row>>
+  {
+    private final Ordering<Row> ordering;
+    private final int limit;
+    private final QueryGranularity granularity;
+
+    public GroupTopNFunction(Ordering<Row> ordering, QueryGranularity granularity, int limit)
+    {
+      this.ordering = ordering;
+      this.granularity = granularity;
+      this.limit = limit;
+    }
+
+    @Override
+    public Sequence<Row> apply(
+        Sequence<Row> input
+    )
+    {
+      Map<Long, MinMaxPriorityQueue<Row>> grouped = Maps.<Long, MinMaxPriorityQueue<Row>>newTreeMap();
+      ArrayList<Row> rows = Sequences.toList(input, Lists.<Row>newArrayList());
+      for (Row row : rows) {
+        long time = granularity.truncate(row.getTimestampFromEpoch());
+        MinMaxPriorityQueue<Row> queue = grouped.get(time);
+        if (queue == null) {
+          grouped.put(time, queue = MinMaxPriorityQueue.orderedBy(ordering).maximumSize(limit).create());
+        }
+        queue.add(row);
+      }
+
+      Iterable<Row> iterable = Collections.emptyList();
+      for (Collection<Row> value : grouped.values()) {
+        iterable = Iterables.concat(iterable, new OrderedPriorityQueueItems<Row>((MinMaxPriorityQueue<Row>) value));
+      }
+      return Sequences.simple(rows.size() > limit ? Iterables.limit(iterable, limit) : iterable);
+    }
+
+      @Override
+    public boolean equals(Object o)
+    {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+
+      GroupTopNFunction that = (GroupTopNFunction) o;
+
+      if (limit != that.limit) {
+        return false;
+      }
+      // todo: it's not comparable
```

Comment:
Throwing exception seemed too harsh. I've changed to `granularity != that.granularity`